File: backend/app/models/notification.py
"""
Notification models and schemas
"""
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class NotificationType(str, Enum):
    """Supported reminder and notification categories."""
    WORKOUT_REMINDER = "workout_reminder"
    MEAL_REMINDER = "meal_reminder"
    WATER_INTAKE_REMINDER = "water_intake_reminder"
    MOTIVATION = "motivation"
    WEEKLY_PROGRESS_REMINDER = "weekly_progress_reminder"


class NotificationPriority(str, Enum):
    """Push priority for reminders."""
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"


class NotificationChannel(str, Enum):
    """Delivery channel for the notification."""
    IN_APP = "in_app"
    PUSH = "push"
    EMAIL = "email"


class NotificationCreate(BaseModel):
    """Schema for creating a notification."""
    type: NotificationType
    title: str = Field(..., min_length=1, max_length=120)
    message: str = Field(..., min_length=1, max_length=500)
    scheduled_for: Optional[datetime] = None
    channel: NotificationChannel = NotificationChannel.IN_APP
    priority: NotificationPriority = NotificationPriority.NORMAL


class NotificationResponse(BaseModel):
    """Response model for a notification record."""
    id: str
    user_id: str
    type: NotificationType
    title: str
    message: str
    channel: NotificationChannel
    priority: NotificationPriority
    status: str
    is_read: bool = False
    scheduled_for: Optional[datetime] = None
    sent_at: Optional[datetime] = None
    read_at: Optional[datetime] = None
    created_at: datetime


class NotificationReadUpdate(BaseModel):
    """Schema for marking a notification as read."""
    is_read: bool = True

File: new_flutter_app/lib/models/notification.dart
/// Notification delivered to the authenticated FitNova user.
class Notification {
  final String id;
  final String userId;
  final String type;
  final String title;
  final String message;
  final String channel;
  final String priority;
  final bool isRead;
  final DateTime? scheduledFor;
  final DateTime createdAt;

  const Notification({
    required this.id,
    required this.userId,
    required this.type,
    required this.title,
    required this.message,
    required this.channel,
    required this.priority,
    required this.isRead,
    this.scheduledFor,
    required this.createdAt,
  });

  factory Notification.fromJson(Map<String, dynamic> json) {
    final data = json['notification'] is Map<String, dynamic>
        ? json['notification'] as Map<String, dynamic>
        : json;
    final createdAt =
        DateTime.tryParse('${data['created_at'] ?? data['createdAt'] ?? ''}') ??
            DateTime.now();

    return Notification(
      id: '${data['id'] ?? data['_id'] ?? ''}',
      userId: '${data['user_id'] ?? data['userId'] ?? ''}',
      type: '${data['type'] ?? 'motivation'}',
      title: '${data['title'] ?? ''}',
      message: '${data['message'] ?? ''}',
      channel: '${data['channel'] ?? 'in_app'}',
      priority: '${data['priority'] ?? 'normal'}',
      isRead: data['is_read'] as bool? ?? data['isRead'] as bool? ?? false,
      scheduledFor: _parseDate(data['scheduled_for'] ?? data['scheduledFor']),
      createdAt: createdAt,
    );
  }

  Map<String, dynamic> toJson() => {
        'type': type,
        'title': title,
        'message': message,
        if (scheduledFor != null)
          'scheduled_for': scheduledFor!.toIso8601String(),
        'channel': channel,
        'priority': priority,
      };

  Notification copyWith({bool? isRead}) => Notification(
        id: id,
        userId: userId,
        type: type,
        title: title,
        message: message,
        channel: channel,
        priority: priority,
        isRead: isRead ?? this.isRead,
        scheduledFor: scheduledFor,
        createdAt: createdAt,
      );

  static DateTime? _parseDate(dynamic value) {
    if (value == null) return null;
    return DateTime.tryParse(value.toString());
  }
}
